password/src/Data/Password/Validate.hs

```diff
@@ -0,0 +1,276 @@
+
+-}
+
+module Data.Password.Validate
+  ( -- * Data type
+    PasswordPolicy (..),
+    InvalidReason (..),
+    CharacterCategory(..),
+    -- * Predicate
+    CharSetPredicate(..),
+    -- * Default values
+    defaultPasswordPolicy,
+    defaultCharSetPredicate,
+    -- * Functions
+    isValidPassword,
+    validatePassword,
+    -- * For internal use
+    isValidPasswordPolicy,
+    isSpecial,
+    defaultCharSet,
+    categoryToPredicate
+  ) where
+
+import Data.Char (chr, isAsciiLower, isAsciiUpper, isDigit, ord)
+import Data.Maybe (catMaybes, fromMaybe)
+import Data.Password.Internal (Password (..))
+#if! MIN_VERSION_base(4,13,0)
+import Data.Semigroup ((<>))
+#endif
+import Data.Text (Text)
+import qualified Data.Text as T
+
+-- $setup
+-- >>> :set -XOverloadedStrings
+--
+-- Import needed libraries.
+--
+-- >>> import Data.Password
+
+-- | Set of policies used to validate 'Password'
+data PasswordPolicy = PasswordPolicy
+    { minimumLength  :: !Int
+    -- ^ Required password minimum length
+    , maximumLength  :: !Int
+    -- ^ Required password maximum length
+    , uppercaseChars :: !(Maybe Int)
+    -- ^ Required number of upper-case characters
+    , lowercaseChars :: !(Maybe Int)
+    -- ^ Required number of lower-case characters
+    , specialChars   :: !(Maybe Int)
+    -- ^ Required number of special characters
+    , digitChars     :: !(Maybe Int)
+    -- ^ Required number of ASCII-digit characters
+    } deriving (Eq, Ord, Show)
+
+-- | Default value for the 'PasswordPolicy'
+defaultPasswordPolicy :: PasswordPolicy
+defaultPasswordPolicy = PasswordPolicy
+  { minimumLength = 8,
+    maximumLength = 32,
+    uppercaseChars = Just 1,
+    lowercaseChars = Just 1,
+    specialChars = Nothing,
+    digitChars = Just 1
+  }
+
+-- | Predicate which defines the characters that can be used for a password.
+newtype CharSetPredicate = CharSetPredicate
+  { getCharSetPredicate :: Char -> Bool
+  }
+
+
+-- | Default character sets consist of uppercase, lowercase letters, numbers,
+-- and special characters
+defaultCharSetPredicate :: CharSetPredicate
+defaultCharSetPredicate =  CharSetPredicate $ \c -> ord c >= 32 && ord c <= 126
+
+-- | Check if given 'Char' is special character
+isSpecial :: Char -> Bool
+isSpecial = \c ->
+    isDefault c && not (or [isAsciiUpper c, isAsciiLower c, isDigit c])
+  where
+    CharSetPredicate isDefault = defaultCharSetPredicate
+
+-- | Character Category
+data CharacterCategory
+  = Uppercase
+  -- ^ Uppercase letters
+  | Lowercase
+  -- ^ Lowercase letters
+  | Special
+  -- ^ Special characters
+  | Digit
+  -- ^ ASCII digits
+  deriving (Eq, Ord, Show)
+
+-- | Convert 'CharacterCategory' into associated predicate function
+categoryToPredicate :: CharacterCategory -> (Char -> Bool)
+categoryToPredicate = \case
+  Uppercase -> isAsciiUpper
+  Lowercase -> isAsciiLower
+  Special -> isSpecial
+  Digit -> isDigit
+
+-- | Possible reason of password being invalid
+data InvalidReason
```

Comment:
Though I agree on not using the `Show` instance to make pretty print error message of `InvalidReason`s, I think it might be a good idea to make a `printInvalidReason :: InvalidReason -> Text` function that turns an `InvalidReason` into a nice error message.

This would be a convenience function for anyone who'd want to display the reason(s) a password is not accepted to the user.

Reply:
This is kind of irrelevant for this PR though, think I'll make a new issue for this after the PR is done.